daemon/algod/api/client/restClient.go

```diff
@@ -72,6 +72,7 @@ type HTTPError struct {
 	StatusCode  int
 	Status      string
 	ErrorString string
+	Data        map[string]any
```

Comment:
What do you think about modifying `HTTPError.Error()` to include the `Data` as serialized json? I like the changes here because they make it easier for our code to inspect the data, but I think it's still important for this info to still be visible to users outside of the code

Reply:
I'm against it, because this is a (potentially) big structure that I don't think belongs in the human readable message, at least not for an unsuspecting client that just does `fmt.Printf("%s", err)`, like `goal`.  I think clients that intend to use this complex info ought to do so explicitly.  It would be very strange, in this example, to have app, pc, opcodes in the first part of the string, and then serialize the JSON too which also contains the app-id and pc. I thought that was the whole reason we put this in here - because we didn't want to change goal output to have a bunch of extra JSON cruft.

With that said, if you think some of the information for a particular goal command ought to be printed, I'd be in favor of those call sites examining the error and printing the info in a nice human way.  (Another PR though?)

Reply:
Good points, especially since this data can get very large/deep with inner txn failures now.

However, there are 2 current uses of the `Data` field which contain useful information for debugging:
1. https://github.com/algorand/go-algorand/blob/eceed7c0d3df0f412ede27c1aa2b68e0fa21ccab/daemon/algod/api/server/v2/handlers.go#L433-L444
2. https://github.com/algorand/go-algorand/blob/eceed7c0d3df0f412ede27c1aa2b68e0fa21ccab/daemon/algod/api/server/v2/handlers.go#L1569-L1576

As I see it, the first one embeds some of the fields into the error message, so losing the `Data` field isn't that bad. It's more of a problem for the second case with boxes.

Since these are the only two cases I could find, I'd be ok with losing the `Data` field when printing if we could augment the error string from the second case to also include the relevant fields (i.e. `total-boxes` and the appropriate limiting variable). What do you think about that?

Reply:
Before I do this, I want to make sure you get what you want.  In the first example, the error message returned to the user is not augmented, only the message logged by `v2.Log.Infof`.  I'm very willing to do the same in the second example.

But I wonder if you want the error message returned to the REST API to contain some of those details?  I'm happy with them being included this new way, and if you are too, I'll just augment the `Infof` of the second example.

If you really want it all in the user string... well, I guess I'll do that too, just let me know whether you think it's worth the effort. By "effort" here, I mean that it's 3 numbers that have some interplay between them, so crafting a good user visible error may be annoying. Noting, of course, that the extra info that is given to REST today is pretty much thrown away by the SDKs, so people do not currently see these numbers.

Reply:
Here's my understanding of the situation:
* You believe clients should not print structured data from errors by default, since it could be large and isn't meant to be human readable
* I'd like for clients to still somehow expose this information because it can contain useful information about an error. Sure, you could reconstruct the request manually and directly perform the HTTP request to see the raw data, but that's a pain and it defeats the purpose of using a client in the first place.

Here are my thoughts on a possible solution:
* The server side behavior for structured error reporting seems correct and useful -- no changes needed there
* When goal receives an error with structured information, it should do the following:
  * By default, only print the message string (structured data is ignored)
  * If somehow a "verbose" option is enabled, print the message string and any structured data that's present. Right now goal routes almost all error printing through `reportErrorf` so we'd probably want to change this or introduce a new variant, since it takes a string arg, not an error, so it would be a bit tedious.

Let me know what you think of this.

For SDKs, I think the error reporting issue matters less. As long as we properly unpack the `Data` field and add that to the produced error object (which likely is not the case right now), it can be inspected by client code easily enough.

I'm going to leave an approval because I think these changes are worth having even without changes to goal, but I'd like to get some alignment on how goal should ideally behave.

Reply:
> If somehow a "verbose" option is enabled, print the message string and any structured data that's present. Right now goal routes almost all error printing through reportErrorf so we'd probably want to change this or introduce a new variant, since it takes a string arg, not an error, so it would be a bit tedious.

Yes, I like this a lot. My thinking is that a "smart" client ought to report more info (even the entire txn failure message, if asked for). It's just that right now, `goal` is a dumb client, and I don't think its lame effort to be smart (just spit out the whole json, including embedded message) cuts it.

I'd like to see that `-v` flag, and when presented it would output something, probably just the `Data` portion of the error, in a structured way.

I'm not totally sure how that `-v should work though.  I could imagine spitting it out after the existing output, but I could also imagine, since it's nice JSON, that a user might prefer an easy way to get _only_ the Data object printed, so it is consumable by tools.  Roughly `--json`.

I'm going to merge this, and I'll think about it more, maybe even open an issue.